cpp/src/lists/contains.cu

```diff
@@ -44,296 +41,241 @@
-  };
+    // If the key is found, return its found position in the list from `found_iter`.
+    return found_iter == end ? NOT_FOUND_SENTINEL : *found_iter;
+  }
 };
 
 /**
- * @brief Functor to search each list row for the specified search keys.
+ * @brief Create a device pointer to the search key(s).
+ *
+ * The returned pointer will be used to construct a validity iterator or an optional iterator for
+ * the keys, which also have the same interface for both `scalar` and `column_device_view`.
+ *
+ * @return Depending on the type of the input key(s), a `scalar` pointer or a `column_device_view`
+ *         pointer will be returned.
  */
-template <bool search_keys_have_nulls>
-struct lookup_functor {
-  template <typename ElementType>
-  struct is_supported {
-    static constexpr bool value =
-      cudf::is_numeric<ElementType>() || cudf::is_chrono<ElementType>() ||
-      cudf::is_fixed_point<ElementType>() || std::is_same_v<ElementType, cudf::string_view>;
-  };
-
-  template <typename ElementType, typename... Args>
-  std::enable_if_t<!is_supported<ElementType>::value, std::unique_ptr<column>> operator()(
-    Args&&...) const
-  {
-    CUDF_FAIL(
-      "List search operations are only supported on numeric types, decimals, chrono types, and "
-      "strings.");
-  }
-
-  std::pair<rmm::device_buffer, size_type> construct_null_mask(
-    lists_column_view const& input_lists,
-    column_view const& result_validity,
-    rmm::cuda_stream_view stream,
-    rmm::mr::device_memory_resource* mr) const
-  {
-    if (!search_keys_have_nulls && !input_lists.has_nulls() && !input_lists.child().has_nulls()) {
-      return {rmm::device_buffer{0, stream, mr}, size_type{0}};
-    } else {
-      return cudf::detail::valid_if(
-        result_validity.begin<bool>(), result_validity.end<bool>(), thrust::identity{}, stream, mr);
-    }
+template <typename SearchKeyType>
+auto get_search_keys_device_view_ptr(SearchKeyType const& search_keys,
+                                     [[maybe_unused]] rmm::cuda_stream_view stream)
+{
+  if constexpr (std::is_same_v<SearchKeyType, cudf::scalar>) {
+    return &search_keys;
+  } else {
+    return column_device_view::create(search_keys, stream);
   }
+}
 
-  template <typename ElementType, typename SearchKeyPairIter>
-  void search_each_list_row(cudf::detail::lists_column_device_view const& d_lists,
-                            SearchKeyPairIter search_key_pair_iter,
-                            duplicate_find_option find_option,
-                            cudf::mutable_column_device_view ret_positions,
-                            cudf::mutable_column_device_view ret_validity,
-                            rmm::cuda_stream_view stream) const
+/**
+ * @brief Dispatch functor to search for key element(s) in the corresponding rows of a lists column.
+ */
+struct dispatch_index_of {
+  template <typename Type,
+            typename SearchKeyType,
+            CUDF_ENABLE_IF(!is_supported_non_nested_type<Type>())>
```

Comment:
I'd prefer to see the valid case defined first, and the failure case defined second.